File: frontend/src/utils/sanitizer.ts
/**
 * XSS protection utilities for frontend
 */

/**
 * Escape HTML special characters to prevent XSS attacks
 */
export function escapeHtml(text: string): string {
  const map: { [key: string]: string } = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#039;',
    '/': '&#x2F;',
  }

  return text.replace(/[&<>"'/]/g, (char) => map[char])
}

/**
 * Remove HTML tags from text
 */
export function stripHtml(html: string): string {
  const tmp = document.createElement('DIV')
  tmp.innerHTML = html
  return tmp.textContent || tmp.innerText || ''
}

/**
 * Sanitize user input by removing dangerous characters
 */
export function sanitizeInput(input: string): string {
  if (!input) return input

  // Remove script tags and their content
  let sanitized = input.replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '')

  // Remove event handlers
  sanitized = sanitized.replace(/on\w+\s*=\s*["'][^"']*["']/gi, '')

  // Remove javascript: protocol
  sanitized = sanitized.replace(/javascript:/gi, '')

  return sanitized
}

/**
 * Validate and sanitize text input with length limit
 */
export function validateTextInput(input: string, maxLength: number): { valid: boolean; sanitized: string; error?: string } {
  if (!input || input.trim().length === 0) {
    return { valid: false, sanitized: '', error: '입력값이 비어있습니다' }
  }

  const trimmed = input.trim()

  if (trimmed.length > maxLength) {
    return { valid: false, sanitized: trimmed, error: `입력값은 ${maxLength}자를 초과할 수 없습니다` }
  }

  const sanitized = sanitizeInput(trimmed)

  return { valid: true, sanitized }
}

File: backend/app/core/sanitizer.py
"""
HTML sanitization utilities for XSS protection
"""
import re
from typing import Optional


def sanitize_html(text: Optional[str]) -> Optional[str]:
    """
    Remove HTML tags and dangerous characters from text to prevent XSS attacks.

    Args:
        text: Input text that may contain HTML tags

    Returns:
        Sanitized text with HTML tags removed
    """
    if not text:
        return text

    # Remove HTML tags
    text = re.sub(r'<[^>]*>', '', text)

    # Remove javascript: protocol
    text = re.sub(r'javascript:', '', text, flags=re.IGNORECASE)

    # Remove on* event handlers
    text = re.sub(r'\bon\w+\s*=', '', text, flags=re.IGNORECASE)

    # Remove data URIs (can be used for XSS)
    text = re.sub(r'data:text/html', '', text, flags=re.IGNORECASE)

    return text


def sanitize_dict(data: dict, fields_to_sanitize: list) -> dict:
    """
    Sanitize specific fields in a dictionary.

    Args:
        data: Dictionary containing data to sanitize
        fields_to_sanitize: List of field names to sanitize

    Returns:
        Dictionary with sanitized fields
    """
    sanitized = data.copy()
    for field in fields_to_sanitize:
        if field in sanitized and isinstance(sanitized[field], str):
            sanitized[field] = sanitize_html(sanitized[field])
    return sanitized
